File: examples/ruby/inheritance.rb
class AbstractList
    def append(element)
        add(size(), element)
    end
end

class ArrayList < AbstractList
    def initialize()
        @array = []
    end

    def add(position, element)
        # Ruby will grow the array automatically
        @array[position] = element
    end

    def size()
        @array.length  # or something
    end
end

students = ArrayList.new

#students.array           # No method called array, only ivar, so this is an error
#students.@array          # This isn’t legal either

# The key difference here is that in Ruby, ivars go in a separate
# object-owned hash table, and don't get resolved up the type chain.

students.size             # Ruby searches students, then ArrayList
students.append("sally")  # Ruby searches students, then ArrayList, then AbstractList

File: examples/python/inheritance.py
class AbstractList:
    def append(self, element):
        self.add(self.size(), element)

class ArrayList(AbstractList):
    def __init__(self):
        self.array = []

    def add(self, position, element):
        if position == self.size():  # Increase array size if needed
            self.array.append(None)
        self.array[position] = element

    def size(self):
        return len(self.array)

students = ArrayList()
students.array            # Python searches students
students.size()           # Python searches students, then ArrayList
students.append("sally")  # Python searches students, then ArrayList, then AbstractList
